### ISSUE TITLE
fs2 vs fs3 - chunks-to-list conversion poor performance

### ISSUE THREAD

**Issue by asjadoun** [typelevel/fs2#2627]
fs2 vs fs3 chunks to list conversion performance

**fs2 version and code:** 
implementation group: 'co.fs2', name: 'fs2-core_2.13', version: '2.5.9'

```
object ChunkToListConversionFs2App extends IOApp {
  private def time[R](action: String, thunk: => R): R = {
    val start = System.currentTimeMillis()
    val result = thunk
    val end = System.currentTimeMillis()
    println(s"$action elapsed time : ${end - start} ms")
    result
  }
  override def run(args: List[String]): IO[ExitCode] = {
    implicit val ec: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
    implicit val timer: Timer[IO] = IO.timer(ExecutionContext.global)
    val stream = Stream
      .repeatEval(IO(System.currentTimeMillis()))
      .chunkN(10000)
      .evalMap(c => IO(time(s"""Using fs2 2.5.9, Chunk of ${c.size} elements conversion 'toList'""", c.toList)))
      .take(3)
      .compile
      .drain
      .unsafeRunSync()

    IO(ExitCode.Success)
  }
}
```

**Output:**
Using fs2 2.5.9, Chunk of 10000 elements conversion 'toList' elapsed time : 5 ms
Using fs2 2.5.9, Chunk of 10000 elements conversion 'toList' elapsed time : 1 ms
Using fs2 2.5.9, Chunk of 10000 elements conversion 'toList' elapsed time : 2 ms

**fs3 version and code:**
implementation group: 'co.fs2', name: 'fs2-core_2.13', version: '3.1.2'

```
object ChunkToListConversionFs3App extends IOApp {
  private def time[R](action: String, thunk: => R): R = {
    val start = System.currentTimeMillis()
    val result = thunk
    val end = System.currentTimeMillis()
    println(s"$action elapsed time : ${end - start} ms")
    result
  }
  override def run(args: List[String]): IO[ExitCode] = {
    implicit val runtime: IORuntime = IORuntime.global

    val stream: Unit = Stream
      .repeatEval(IO(System.currentTimeMillis()))
      .chunkN(10000)
      .evalMap(c => IO(time(s"""Using fs2 3.1.2, Chunk of ${c.size} elements conversion 'toList'""", c.toList)))
      .take(3)
      .compile
      .drain
      .unsafeRunSync()

    IO(ExitCode.Success)
  }}
```

**Output:**
Using fs2 3.1.2, Chunk of 10000 elements conversion 'toList' elapsed time : 1518 ms
Using fs2 3.1.2, Chunk of 10000 elements conversion 'toList' elapsed time : 1429 ms
Using fs2 3.1.2, Chunk of 10000 elements conversion 'toList' elapsed time : 1455 ms

**FS3 is nearly 1000 times slower than fs2**

**Comment by mpilquist** [comment]
Thanks for reporting -- I think this is the same issue as described in #2603. In your fs2 3 example, what happens if you change to `c.compact.toList`?

**Comment by asjadoun** [comment]
It does help and reduced the time to < 10 ms. However, it is no as good as compare to fs 2.5.9. In our application, chunks can be from 10,000 elements to 1,000,000 elements of up to 1 - 3 kb each in size.

Just for test, I ran both fs2 and fs3 code with 1M chunk size -

```
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 98 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 100 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 420 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 256 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 92 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 77 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 75 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 110 ms
Using fs2 3.1.2, Chunk of 1000000 elements conversion 'toList' elapsed time : 521 ms
```
vs
```
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 22 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 16 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 6 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 7 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 5 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 5 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 15 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 6 ms
Using fs2 2.5.9, Chunk of 1000000 elements conversion 'toList' elapsed time : 6 ms
```

fs2 perform far better than fs3 with compact.

**Comment by mpilquist** [comment]
Directly comparing the 2.5 `.toList` with the 3.x `.compact.toList` is problematic because 2.5 pays the compaction cost in `chunkN`. It would be better / more accurate to use this for 3.x: `.chunkN.map(_.compact).evalMap(...)`. Basically, don't include the compaction cost in the benchmark, or if you want to, include it for both 2.5 and 3.x.

The main issue here is caused by `chunkN` (and the lower level `pull.unconsN`) emitting a `Chunk.Queue` now. The `unconsN` operation works by enqueuing each chunk in to a `Chunk.Queue` until the desired size is reached. In 2.5, `Chunk.Queue` was not a subtype of `Chunk` so we'd convert the `Chunk.Queue` to a `Chunk`, compacting in the process. In 3.x, `Chunk.Queue` is a subtype of `Chunk` so it's emitted directly without compaction. Adding a `.compact` should restore the 2.5 performance.

By making `Chunk.Queue` a subtype of `Chunk`, we're able to avoid all sorts of unnecessary copying/compaction. However, the tradeoff is that your average constituent chunk size approaches 1 (i.e. number of chunks approaches the number of total elements), then we get O(n) indexed lookup.

The quick fix is simply making `.unconsN` compact, which would restore the 2.5 performance characteristics. There are lots of other uses of `Chunk.Queue` though, and most of them are susceptible to the same issue. Hence, I'm wondering if we can do something smarter here, like having `Chunk.Queue` auto-compact if ratio of size to chunk count reaches a certain threshold.

**Comment by mpilquist** [comment]
I tried to implement auto-compaction upon first call to `apply` and realized now why we can't do this. The issue is that in all of these contexts -- e.g., `chunkN` or the implementation of `Chunk.Queue#apply` -- we don't have access to a `ClassTag[O]`.

**Comment by mpilquist** [comment]
Fixed in 3.1.3
